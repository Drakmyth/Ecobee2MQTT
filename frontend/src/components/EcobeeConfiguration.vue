<template>
    <div class="ecobee-config">
        <h2>Ecobee Configuration</h2>
        <form>
            <label for="ecobee-appkey">App Key:</label>
            <input type="text" id="ecobee-appkey" name="ecobee-appkey" v-model="appkey" /><br />
            <button @click.prevent=requestPin>Request PIN</button><br />
            <span>PIN: {{ pin }}</span><br />
            <div v-if="authStatus === 'pending'">
                <span v-if="timeremaining > 0">Time Remaining: {{ formatTime(timeremaining) }}</span><br />
                <span v-if="timeremaining <= 0">PIN expired. Please request a new PIN.</span><br />
            </div>
            <span v-if="authStatus === 'authorized'">Account Authorized!</span>
        </form>
    </div>
</template>

<script setup lang="ts">
import axios from 'axios';
import { Ref, ref } from 'vue';
import { io } from 'socket.io-client';

type AuthStatus = 'none' | 'pending' | 'authorized';

const appkey = ref('');
const pin = ref('');
const timeremaining = ref(0);
let timer: number;
let authStatus: Ref<AuthStatus> = ref('none');

const socket = io('ws://localhost:3001');

socket.on('authorized', (arg) => {
    authStatus.value = 'authorized';
    console.log(arg);
});

const requestPin = () => {
    axios.post('http://localhost:3001/api/authorization', {
        appkey: appkey.value,
        scope: 'smartWrite'
    }).then(resp => {
        console.log(resp.data);
        pin.value = resp.data.pin;
        authStatus.value = 'pending';
        timeremaining.value = resp.data.expires_in;
        timer = setInterval(onTimerTick, 1000);
    }).catch(error => {
        console.error(error);
    });
};

const onTimerTick = () => {
    timeremaining.value -= 1;
    if (timeremaining.value <= 0) {
        clearInterval(timer);
    }
};

const formatTime = (seconds: number) => {
    return Math.floor(seconds / 60) + ':' + ('0' + Math.floor(seconds % 60)).slice(-2);
};

</script>

<style>
</style>
