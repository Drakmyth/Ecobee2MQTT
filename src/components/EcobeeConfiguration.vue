<script setup lang="ts">
</script>

<template>
    <div class="ecobee-config">
        <h2>Ecobee Configuration</h2>
        <form>
            <label for="apikey">API Key:</label>
            <input type="text" id="apikey" name="apikey" /><br />
            <button>Request PIN</button><br />
            <span>PIN: </span>
            <span>ABC123</span><br />
            <span>Time Remaining: </span>
            <span>9:32</span><br />
            <button>Complete Authorization</button>
        </form>
    </div>
</template>

<style>
</style>
