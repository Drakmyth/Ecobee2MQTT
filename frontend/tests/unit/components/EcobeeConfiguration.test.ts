import { shallowMount } from '@vue/test-utils';
import EcobeeConfiguration from '@/components/EcobeeConfiguration.vue';

describe('EcobeeConfiguration.vue', () => {
    test('renders correctly', () => {
        const wrapper = shallowMount(EcobeeConfiguration);
        expect(wrapper.html()).toMatchSnapshot();
    });
});
