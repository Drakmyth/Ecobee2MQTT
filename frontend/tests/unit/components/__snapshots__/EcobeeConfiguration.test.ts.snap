// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcobeeConfiguration.vue renders correctly 1`] = `
"<div class=\\"ecobee-config\\">
  <h2>Ecobee Configuration</h2>
  <form><label for=\\"ecobee-appkey\\">App Key:</label><input type=\\"text\\" id=\\"ecobee-appkey\\" name=\\"ecobee-appkey\\"><br>
    <div><button>Request PIN</button><br><span>PIN: </span><br></div>
    <!--v-if-->
    <!--v-if-->
  </form>
</div>"
`;
